capsules/src/rng.rs

```diff
@@ -0,0 +1,177 @@
+
+                            // 3. Zip over the randomness iterator and chunks
+                            //    of up to 4 bytes from the buffer.
+                            for (inp, outs) in randomness.take(remaining_ints)
+                                .zip(buf.chunks_mut(4)) {
+
+                                // 4. For each word of randomness input, update
+                                //    the remaining and idx and add to buffer.
+                                for (i, b) in outs.iter_mut().enumerate() {
+                                    *b = ((inp >> i * 8) & 0xff) as u8;
+                                    app.remaining -= 1;
+                                    app.idx += 1;
+                                }
+                            }
+                        }
+
+                        // Replace taken buffer
+                        app.buffer = Some(buffer);
+
+                    });
+
+                    if app.remaining > 0 {
+                        done = false;
+                    } else {
+                        app.callback.map(|mut cb| {
+                            cb.schedule(0, app.idx, 0);
+                        });
+                    }
+                }
+            });
+
+            // Check if done switched to false. If it did, then that app
+            // didn't get enough random, so there's no way there is more for
+            // other apps.
+            if done == false {
+                break;
+            }
+        }
+
+        if done {
+            self.getting_randomness.set(false);
+            rng::Continue::Done
+        } else {
+            rng::Continue::More
+        }
+    }
+}
+
+impl<'a, RNG: rng::RNG> Driver for SimpleRng<'a, RNG> {
+    fn allow(&self, appid: AppId, allow_num: usize, slice: AppSlice<Shared, u8>) -> isize {
+        // pass buffer in from application
+        match allow_num {
+            0 => {
+                self.apps
+                    .enter(appid, |app, _| {
+                        app.buffer = Some(slice);
+                        0
+                    })
+                    .unwrap_or(-1)
+            }
+            _ => -1,
+        }
+    }
+
+    fn subscribe(&self, subscribe_num: usize, callback: Callback) -> isize {
+        match subscribe_num {
+            0 => {
+                self.apps
+                    .enter(callback.app_id(), |app, _| {
+                        app.callback = Some(callback);
+                        0
+                    })
+                    .unwrap_or(-1)
+            }
+
+            // default
+            _ => -1,
+        }
+    }
+
+    fn command(&self, command_num: usize, data: usize, appid: AppId) -> isize {
+        match command_num {
+            0 => /* Check if exists */ 0,
+
+            // Ask for a given number of random bytes.
+            1 => {
+                self.apps
+                    .enter(appid, |app, _| {
+                        app.remaining = data;
```

Comment:
Shouldn't you be limiting them here to only:
1) ask for bytes if there is a buffer allowed already
2) ask for a number of bytes up to the buffer's size
3) possibly, only ask for bytes if there is a subscribe already